database/src/cache.rs

```diff
@@ -1,75 +1,214 @@
+    /// Indicates if this cache was set to be tracked.
+    tracked: bool,
+    /// The max size of this cache. Units (bytes or some logical unit) depend on
+    /// caller logic and the implementation of `MemSizeEstimator`.
+    max_size: usize,
+    /// Min units to keep in the cache even if passing tracked size limit.
+    min_units: usize,
+    /// Indicates whether tracking is in bytes mode
+    mem_mode: MemMode,
 }
 
-impl<TKey: Clone + std::hash::Hash + Eq + Send + Sync, TData: Clone + Send + Sync, S: BuildHasher + Default> Cache<TKey, TData, S> {
-    pub fn new(size: u64) -> Self {
-        Self { map: Arc::new(RwLock::new(IndexMap::with_capacity_and_hasher(size as usize, S::default()))), size: size as usize }
+impl From<CachePolicy> for CachePolicyInner {
+    fn from(policy: CachePolicy) -> Self {
+        match policy {
+            CachePolicy::Empty => CachePolicyInner { tracked: false, max_size: 0, min_units: 0, mem_mode: MemMode::Units },
+            CachePolicy::Unit(max_size) => CachePolicyInner { tracked: false, max_size, min_units: 0, mem_mode: MemMode::Units },
+            CachePolicy::Tracked(max_size, mem_mode) => CachePolicyInner { tracked: true, max_size, min_units: 0, mem_mode },
+            CachePolicy::LowerBoundedTracked { max_size, min_units, mem_mode } => {
+                CachePolicyInner { tracked: true, max_size, min_units, mem_mode }
+            }
+        }
+    }
+}
+
+struct Inner<TKey, TData, S = RandomState>
+where
+    TKey: Clone + std::hash::Hash + Eq + Send + Sync,
+    TData: Clone + Send + Sync + MemSizeEstimator,
+{
+    // We use IndexMap and not HashMap because it makes it cheaper to remove a random element when the cache is full.
+    map: IndexMap<TKey, TData, S>,
+    tracked_size: usize,
+}
+
+impl<TKey, TData, S> Inner<TKey, TData, S>
+where
+    TKey: Clone + std::hash::Hash + Eq + Send + Sync,
+    TData: Clone + Send + Sync + MemSizeEstimator,
+    S: BuildHasher + Default,
+{
+    /// Evicts items until meeting cache policy requirements
+    fn evict(&mut self, policy: &CachePolicyInner) {
+        // We allow passing tracked size limit as long as there are no more than min_units units
+        while self.tracked_size > policy.max_size && self.map.len() > policy.min_units {
+            if let Some((_, v)) = self.map.swap_remove_index(rand::thread_rng().gen_range(0..self.map.len())) {
+                self.tracked_size -= v.estimate_size(policy.mem_mode)
+            }
+        }
+    }
+
+    fn insert(&mut self, policy: &CachePolicyInner, key: TKey, data: TData) {
+        if policy.tracked {
+            let new_data_size = data.estimate_size(policy.mem_mode);
+            self.tracked_size += new_data_size;
+            if let Some(removed) = self.map.insert(key, data) {
+                self.tracked_size -= removed.estimate_size(policy.mem_mode);
+            }
+            self.evict(policy);
+        } else {
+            if self.map.len() == policy.max_size {
```

Comment:
I added a `MemMode::Undefined` and renamed `CachePolicy::Units` to `CachePolicy::Count` to avoid confusion. Non tracking policy means by definition that we don't care about the memory mode

Reply:
Re `self.map.len() > policy.max_size` the current logic is tight in the sense that it never passes the limit (bcs we check and evict before insertion). If we would change to `>=` then we also need a `while` loop. To me it seems more safe to keep the logic tight since this code should rarely change. Wdyt @someone235 ?

Reply:
I agree that we should keep it as is to avoid the `while`